



# Homework 1 Part 1: Pandas

Today, we'll practice data exploration in pandas! Each cell should consist of **a single line of pandas code**, answering the question, unless mentioned that **multiple lines of pandas can be used**. Also, **Instructions are provided before each coding exercise; please follow them accordingly**.

First, download the "Pizza Sales Dataset" from Kaggle.com (https://www.kaggle.com/datasets/nextmillionaire/pizza-sales-dataset?resource=download) and place it in this directory. **Make sure to create a Kaggle account first**.


List of some helpful functions in general:


* read_csv
* head
* unique
* sum
* mean
* median
* groupby
* apply (An important note about this one--pay careful attention to the weird axis argument. When you apply over a series, you often don't need it, but when you apply over a dataframe axis=1 and axis=0 will do very different things.)
* value_counts
* df.columns ('columns' is a dataframe variable that tracks the columns) [note: it is not a function, rather an attribute]
* isin
* fillna
* astype
* hist

## Section A: The Basics Setup (TOTAL POINTS: 04)

***Remember, all answers must be in A SINGLE LINE OF CODE in this section A.***

**(1 POINT)** First, import the pandas library as "pd"

In [2]:
import pandas as pd

**(1 POINT)** Now, read the dataframe and store it in a variable called "df".

In [3]:
df = pd.read_csv("./pizza_sales.csv")

**(1 POINT)** Let's get a feel for our dataframe. Print out the list of columns


In [4]:
df.columns

Index(['pizza_id', 'order_id', 'pizza_name_id', 'quantity', 'order_date',
       'order_time', 'unit_price', 'total_price', 'pizza_size',
       'pizza_category', 'pizza_ingredients', 'pizza_name'],
      dtype='object')

**(1 POINT)** Now print out the first ten elements. There's a single function that does it by default.

In [5]:
df.head(10)

pizza_id  order_id  pizza_name_id  quantity order_date order_time  \
0       1.0       1.0     hawaiian_m       1.0   1/1/2015   11:38:36   
1       2.0       2.0  classic_dlx_m       1.0   1/1/2015   11:57:40   
2       3.0       2.0  five_cheese_l       1.0   1/1/2015   11:57:40   
3       4.0       2.0    ital_supr_l       1.0   1/1/2015   11:57:40   
4       5.0       2.0     mexicana_m       1.0   1/1/2015   11:57:40   
5       6.0       2.0     thai_ckn_l       1.0   1/1/2015   11:57:40   
6       7.0       3.0    ital_supr_m       1.0   1/1/2015   12:12:28   
7       8.0       3.0   prsc_argla_l       1.0   1/1/2015   12:12:28   
8       9.0       4.0    ital_supr_m       1.0   1/1/2015   12:16:31   
9      10.0       5.0    ital_supr_m       1.0   1/1/2015   12:21:30   

   unit_price  total_price pizza_size pizza_category  \
0       13.25        13.25          M        Classic   
1       16.00        16.00          M        Classic   
2       18.50        18.50          L         Veggie   
3       20.75        20.75          L        Supreme   
4       16.00        16.00          M         Veggie   
5       20.75        20.75          L        Chicken   
6       16.50        16.50          M        Supreme   
7       20.75        20.75          L        Supreme   
8       16.50        16.50          M        Supreme   
9       16.50        16.50          M        Supreme   

                                   pizza_ingredients  \
0           Sliced Ham, Pineapple, Mozzarella Cheese   
1  Pepperoni, Mushrooms, Red Onions, Red Peppers,...   
2  Mozzarella Cheese, Provolone Cheese, Smoked Go...   
3  Calabrese Salami, Capocollo, Tomatoes, Red Oni...   
4  Tomatoes, Red Peppers, Jalapeno Peppers, Red O...   
5  Chicken, Pineapple, Tomatoes, Red Peppers, Tha...   
6  Calabrese Salami, Capocollo, Tomatoes, Red Oni...   
7  Prosciutto di San Daniele, Arugula, Mozzarella...   
8  Calabrese Salami, Capocollo, Tomatoes, Red Oni...   
9  Calabrese Salami, Capocollo, Tomatoes, Red Oni...   

                         pizza_name  
0                The Hawaiian Pizza  
1          The Classic Deluxe Pizza  
2             The Five Cheese Pizza  
3         The Italian Supreme Pizza  
4                The Mexicana Pizza  
5            The Thai Chicken Pizza  
6         The Italian Supreme Pizza  
7  The Prosciutto and Arugula Pizza  
8         The Italian Supreme Pizza  
9         The Italian Supreme Pizza

## Section B: Exploration Part 01 (TOTAl POINTS:14)
***Remember, all answers must be in A SINGLE LINE OF CODE in this section B.***

**(2 POINTS)** How many pizza were orderd in **total**?

In [7]:
df.shape[0]

48620

**(2 POINTS)** Print out what is the **Median** of **total price**?


In [10]:
df["total_price"].median()

16.5

**(3 POINTS)** Making it a bit more complicated — print out the median of the **'total price'** but only for the pizza_name_id named **'five_cheese_l'**

In [15]:
df.loc[df.pizza_name_id == "five_cheese_l"]["total_price"].median()

18.5

**(2 POINTS)** Calculate the **mean 'unit price'** for each unique **'pizza name id'**

Sample answer could be: <br>

bbq_ckn_l        20.75<br>
bbq_ckn_s        12.75
etc.

In [18]:
df.groupby("pizza_name_id").apply(lambda x: x["unit_price"].mean())

pizza_name_id
bbq_ckn_l        20.75
bbq_ckn_m        16.75
bbq_ckn_s        12.75
big_meat_s       12.00
brie_carre_s     23.65
                 ...  
the_greek_xl     25.50
the_greek_xxl    35.95
veggie_veg_l     20.25
veggie_veg_m     16.00
veggie_veg_s     12.00
Length: 91, dtype: float64

**(2 POINTS)** Print out the number of pizza by **'pizza size'**.

Your results should look something like: <br>
L      18526 <br>
M      15385 <br>
etc.

In [19]:
df.groupby("pizza_size").apply(lambda x: x.shape[0])

pizza_size
L      18526
M      15385
S      14137
XL       544
XXL       28
dtype: int64

**(1 POINT)** Print out the number of pizza by **'pizza name'**..

Your results should look something like: <br>
The Classic Deluxe Pizza                      2416 <br>
The Barbecue Chicken Pizza                    2372 <br>
etc.

In [20]:
df.groupby("pizza_name").apply(lambda x: x.shape[0])

pizza_name
The Barbecue Chicken Pizza                    2372
The Big Meat Pizza                            1811
The Brie Carre Pizza                           480
The Calabrese Pizza                            927
The California Chicken Pizza                  2302
The Chicken Alfredo Pizza                      980
The Chicken Pesto Pizza                        961
The Classic Deluxe Pizza                      2416
The Five Cheese Pizza                         1359
The Four Cheese Pizza                         1850
The Greek Pizza                               1406
The Green Garden Pizza                         987
The Hawaiian Pizza                            2370
The Italian Capocollo Pizza                   1414
The Italian Supreme Pizza                     1849
The Italian Vegetables Pizza                   975
The Mediterranean Pizza                        923
The Mexicana Pizza                            1456
The Napolitana Pizza                          1451
The Pepper Salami Pi

**(2 POINTS)** Print just the line for order date "1/1/2015" and order time "12:21:30". It will print/display **only one row**!

In [21]:
df.loc[(df.order_date == "1/1/2015") & (df.order_time == "12:21:30")]

pizza_id  order_id pizza_name_id  quantity order_date order_time  \
9      10.0       5.0   ital_supr_m       1.0   1/1/2015   12:21:30   

   unit_price  total_price pizza_size pizza_category  \
9        16.5         16.5          M        Supreme   

                                   pizza_ingredients  \
9  Calabrese Salami, Capocollo, Tomatoes, Red Oni...   

                  pizza_name  
9  The Italian Supreme Pizza

## Section C: Exploration Part 02 : Little bit advanced (TOTAL POINTS: 07)
***In this section C, you can use MULTIPLE LINE OF ANSWERS if needed***

**(3 POINTS)** Q1: Part one: Create a series 'c' by summing the 'quantity' for each unique 'pizza_name'. Then, sort the series 'c' in descending order, storing the result in 'c_sorted'." **hints**: use "sort_values"

**(1 POINT)** Q1: Part2: Now from "c_sorted";  find out which was the most-ordered pizza? **hints: it will be just one line code**

**(3 POINTS)** Display How many **Classic** pizza were ordered in **29-10-2015**

## Section D: Modifications : (TOTAL POINTS: 07)
***Remember, all answers must be in A SINGLE LINE OF CODE in this section D.***

Let's start modifying our dataframe! Remember, dataframe operations return a copy by default, so you'll either need to use the inplace=True, or just assign the dataframe back into itself (as in, df = df.someFunction()).

**(2 POINTS)** Take the Pizza Size column and **replace** 'M' with 'Medium' and 'L' with 'Large' in the 'pizza_size' column of the 'df' dataframe. Display the *first 10 rows* of the updated dataframe."

**(1 POINT)** Now make a new 'revenue' column, which should be total price + 10 dollars, as the company accidentally factored in an old discount.

**(2 POINTS)** Convert the 'order_date' datetime format using "to_datetime()". Display the updated 'order_date' column.

**(2 POINTS)** Calculate the total revenue for the period from January 1, 2015, to February 28, 2015

## Section E: Graph Visualization : (TOTAL POINTS: 02)
***Remember, all answers must be in A SINGLE LINE OF CODE in this section E. ***

**(2 POINTS)** Let's do some very basic graphing here! Create a histogram for the 'order_date' column. You can set color on your own choice.

Create a histogram that visualizes the frequency of each pizza category. Use the 'pizza_category' column from the dataset.

# ADDITIONAL Bonus Section (TOTAL POINTS: 06)
**Multiple lines code are allowed in this section**

**(5 POINTS)**

Now you need to load another dataset named "costs_ingredients.csv" in the directory.

The company forgot to calculate their profit! Recall that profit is revenue minus total **costs** to the company. You are given a list below of the integredient costs. Using these costs, compute first the total cost of each order, and then the respective revenue. **Hints: You may need to do some python code  with pandas**

**(1 POINT)** Plot a histogram of the profit. Do you see a problem... ?

Explain in a text block below.

Write down the explaination: the average is below zero; so not making any money